torch/compiler/__init__.py

```diff
@@ -381,3 +383,30 @@ def is_dynamo_compiling() -> bool:
         >>>     # ...rest of the function...
     """
     return False
+
+
+@contextlib.contextmanager
+def skip_guard_eval_unsafe() -> Generator[None, None, None]:
+    """
+    CAUTION - This API is unsafe and should only be used if your setup meets the
+    following conditions.
+
+    torch.compile uses a guard system to support recompilations and choose which
+    compiled artifact to run at runtime.  These guards, though efficient, add
+    some overhead, which may impact performance in scenarios where you need to
+    optimize for minimal guard processing time.
+
+    This API enables you to disable guard evaluation, assuming that you have
+    warmed up the compiled model with a sufficient variety of inputs. This
+    assumption means that, after the warmup phase, no further recompilations
+    will be necessary.
+
+    If this assumption fails, there is a risk of silently producing incorrect
+    results (hence the term "unsafe" in the API name).
```

Comment:
There is some sort of nontrivial proof of correctness here that relies on the "trie-ness" of guard sets.

In particular, consider this near-miss counterexample. I have two compiled code with these guards:

```
a1 and a2
b1
```

Let us suppose I triggered compilation of these frames with:

```
a1=True, a2=True, b1=False
a1=False, a2=False, b1=True
```

These inputs will trigger compilation of both frames. From these two compilations, a1 will have a nonzero fail count.

When we skip guard evaluation, our adjusted guards are now:

```
a1
True
```

But this is now unsound EVEN if the user only needs to run the compiled code above. Specifically, if I have `a1=True, a2=False, b1=True`, the second graph would hit (since b1 == True), but because a1 is True I will use the first compiled code anyway.

Why is this a near-miss counterexample? Assume guards are never elided, it is impossible for the second frame to have only b1 as its guard. Instead, our guards should be:

```
a1 and a2
not a1 and b1
```

In other words, the invariant is that every compiled code must have the EXACTLY same guards, up until a "critical" guard which must be included in both guard sets but then must vary before you're allowed to now have different guards. Under these guards, `a1=True, a2=False, b1=True` isn't covered by either compiled code and so under the contract here it is OK to drop.

Unfortunately, it's difficult for me to tell if we 100% can rely on this invariant. One easy way to get into trouble is if we sometimes elide guards (which we do!) I also have some vague concern with what happens when we reorder guards but maybe it's OK because you always have to accumulate a failure on the critical guard before anything gets reordered, so the critical guards won't get dropped.

Reply:
Yeah, this does not work 😿 

I have to think more.